terraform/modules/iam/iam.tf

```diff
@@ -9,20 +9,63 @@ locals {
+
+  onm_compartment_scope = var.root_compartment_ocid == var.oci_onm_compartment_ocid ? "tenancy" : "compartment id ${var.oci_onm_compartment_ocid}"
+  oke_compartment_scope = var.root_compartment_ocid == var.oke_compartment_ocid ? "tenancy" : "compartment id ${var.oci_onm_compartment_ocid}"
+
+  policy_stmts = {
+    metric_upload = ["Allow dynamic-group ${local.dynamic_group_name} to use METRICS in ${local.onm_compartment_scope} WHERE target.metrics.namespace = 'mgmtagent_kubernetes_metrics'"],
+    log_upload    = ["Allow dynamic-group ${local.dynamic_group_name} to {LOG_ANALYTICS_LOG_GROUP_UPLOAD_LOGS} in ${local.onm_compartment_scope}"],
+    discovery_api = ["Allow dynamic-group ${local.dynamic_group_name} to {LOG_ANALYTICS_DISCOVERY_UPLOAD} in tenancy"],
+    tag_namespace = ["Allow dynamic-group ${local.dynamic_group_name} to use tag-namespaces in tenancy ${local.tags_policy_where_clause}"]
+    infra_discovery_stmt = [
+      "Allow dynamic-group ${local.dynamic_group_name} to inspect compartments in tenancy",
+      # https://docs.oracle.com/en-us/iaas/Content/Identity/Reference/contengpolicyreference.htm
+      "Allow dynamic-group ${local.dynamic_group_name} to use clusters in tenancy where target.cluster.id=${var.oke_cluster_ocid}",
+      "Allow dynamic-group ${local.dynamic_group_name} to read cluster-node-pools in tenancy",
+      # https://docs.oracle.com/en-us/iaas/Content/Identity/Reference/corepolicyreference.htm
+      # Note:
+      # Customers will need to create additional policies to support VCN and subnets in non-OKE compartments
+      "Allow dynamic-group ${local.dynamic_group_name} to read vcns in ${local.oke_compartment_scope}",
+      "Allow dynamic-group ${local.dynamic_group_name} to use subnets in ${local.oke_compartment_scope}",
+      # https://docs.oracle.com/en-us/iaas/Content/Identity/Reference/lbpolicyreference.htm
+      "Allow dynamic-group ${local.dynamic_group_name} to use load-balancers in ${local.oke_compartment_scope}"
+    ],
+    # Note:
+    # In Order to read data from an existing log-group (which can we part of any compartment),
+    # We must allow read access in at least, both ONM and OKE compartments
+    # Compartment of Logging LogGroup is not known at the time of policy creation via stack
+    # We assume that Logging Log Groups are only created in either OKE or ONM compartments
+    service_discovery_stmt = var.create_service_discovery_policies ? distinct([
+      "Allow dynamic-group ${local.dynamic_group_name} to manage log-groups in ${local.onm_compartment_scope}",
+      "Allow dynamic-group ${local.dynamic_group_name} to manage log-groups in ${local.oke_compartment_scope}",
+
+      "Allow dynamic-group ${local.dynamic_group_name} to read log-content in ${local.onm_compartment_scope}",
```

Comment:
lets put a where clause to restrict this policy to service connector resource if that is sufficient